### ISSUE TITLE
🐛 [BUG] Nondeterminism in FullyConnectedTensorProduct

### ISSUE THREAD

**Issue by tomato1mule** [e3nn/e3nn#352]
**Describe the bug**

The outputs of o3.FullyConnectedTensorProduct are sometimes slightly different even when I run exactly the same code and with same seed.

The difference is very small (order of 1e-7 ~ 1e-10) but is non-negligible as the error accumulates.

The outputs are fully deterministic in the same run of the code, so I presume the nondeterminism is coming from the initialization of the e3nn package.

I checked the state_dicts but they are exactly identical. 

The nondeterminism occurs even when the TensorProduct module should be stateless (shared_weights=False, internal_weights=False).

I am currently pickling all the TensorProduct modules to avoid this nondeterminism.

It would be nice if this nondeterminism is resolved in the next release.

**To Reproduce**

```Python
seed = 0
import random
import numpy as np
import torch

random.seed(seed)
np.random.seed(seed)
torch.manual_seed(seed)
torch.use_deterministic_algorithms(True)
torch.set_printoptions(precision=20, sci_mode=False)

import e3nn.o3 as o3


irreps = o3.Irreps('10x0e+10x1e+4x2e+2x3e')
irreps_sh = o3.Irreps.spherical_harmonics(lmax=3, p=1)

tp = o3.FullyConnectedTensorProduct(irreps, irreps_sh, irreps, shared_weights=False, internal_weights=False)

N_batch = 100
inputs = irreps.randn(N_batch,-1)
sh = irreps_sh.randn(N_batch,-1)
weights = torch.randn(N_batch, tp.weight_numel)
outputs = tp(inputs, sh, weights)

if False:      #   Please set this flag to True to save pickles. Once they are saved, set the flag to False and restart the code to see the difference.
    torch.save(outputs, 'outputs.pk')
    torch.save(tp.state_dict(), 'state_dict.pk')
    import gzip
    import pickle
    with gzip.open("tp.gzip",'wb') as f:
        pickle.dump(tp, f)
else:
    outputs_pk = torch.load('outputs.pk')
    state_dict_pk = torch.load('state_dict.pk')
    import gzip
    import pickle
    with gzip.open("tp.gzip",'rb') as f:
        tp_pk = pickle.load(f)

    print((outputs == outputs_pk).all())
    print((outputs != outputs_pk).nonzero())
```

**Expected behavior**

The output should be like
```
tensor(True)
tensor([], size=(0, 2), dtype=torch.int64)
```
However, with ~50% chance, the output is like:

```
tensor(False)
tensor([[ 0, 15],
        [ 0, 16],
        [ 0, 18],
        ...,
        [99, 65],
        [99, 68],
        [99, 69]])
```

**Environment (please complete the following information):**
 - OS: Ubuntu 20.04, 22.04
 - python 3.9.12
 - python environment (commands are given for python interpreter):
   - e3nn version: 0.4.4, 0.5.0
   - pytorch version: 1.11.0+cu113
   - pytorch_geometric version: No module named 'pytorch_geometric'
 - (if relevant) GPU support with CUDA
   -  The nondeterminism can be reproduced with cpu mode.

**Screenshot**

![Screenshot from 2022-08-07 21-46-37](https://user-images.githubusercontent.com/45463512/183291230-c119a868-0d54-485c-ad95-f6ad6ca782b5.png)

Occasionally, the outputs of the o3.FullyConnectedTensorProduct are slightly (order of 1e-7 ~ 1e-10) different even when I run exactly the same code.

**Additional context**
It seems like there are two different modes in the initialization.
With around 50% change, the output is the same and with anothor 50% chance, the output is slightly different but the difference is always the same.

**Comment by mariogeiger** [comment]
Hi, interesting bug.
I managed to reproduce and I observed that if I set `_optimize_einsums=False` it becomes deterministic.
@Linux-cpp-lisp do you have any idea?

**Comment by Linux-cpp-lisp** [comment]
Huh, interesting...

I thought that the einsum optimization should be fully deterministic because I default to "optimal" strategy which does an exhaustive search of contraction orders:
 - https://github.com/Linux-cpp-lisp/opt_einsum_fx/blob/main/opt_einsum_fx/_opt_ein.py#L103
 - https://optimized-einsum.readthedocs.io/en/stable/optimal_path.html

I think in your case there are at least two different optimal contraction orders (since the `i` and `k` indices have the same irreps and dim). The large number of `set` / `dict` operations in `opt_einsum`'s [implementation of the `"optimal"` strategy](https://optimized-einsum.readthedocs.io/en/stable/_modules/opt_einsum/paths.html#optimal) make me suspect that which optimal order gets picked depends on some Python hash ordering. This would also explain the ~50% incidence of the difference.

To test this, can you check if the output is deterministic when you also fix `PYTHONHASHSEED`:
```bash
$ PYTHONHASHSEED=0 python myscript.py
```

Also, make sure that your `opt_einsum` (not `opt_einsum_fx`) is upgraded, they may in the past have had some nondeterminism problems? (https://github.com/dgasmith/opt_einsum/issues/62, for example)

**Comment by tomato1mule** [comment]
Many thanks to your replies.

Manually setting PYTHONHASHSEED turned out to be the solution.

After manually setting it, the outputs are now completely deterministic.

**Comment by Linux-cpp-lisp** [comment]
@tomato1mule Great! Glad to hear.

This is something that could be worth adding a note to `opt_einsum_fx` (happy to accept a PR) and possibly file an "issue" (even though it's not a bug) with a minimal repro on `opt_einsum` itself...
